decompiler/pipeline/controlflowanalysis/restructuring_commons/condition_based_refinement.py

```diff
@@ -2,14 +2,164 @@
+
+    def _remove_formula(self, formula: Formula) -> Iterator[Union[Clause, Symbol]]:
+        """Remove the given formula from the graph."""
+        yield from self._logic_graph.successors(formula)
+        self._logic_graph.remove_node(formula)
+        for symbol in self._symbols_of_formula[formula]:
+            yield from self._remove_symbol_from_formula(formula, symbol)
+        del self._candidates[formula.ast_node]
+
+    def _remove_clause(self, clause: Clause) -> Iterator[Union[Formula, Symbol]]:
+        """Remove the given clause from the graph."""
+        if clause.formula in self._logic_graph:
+            if self._logic_graph.out_degree(clause.formula) == 1:
+                yield clause.formula
+            else:
+                for symbol in (s for s in self._logic_graph.successors(clause) if not has_path(self._logic_graph, clause.formula, s)):
+                    self._symbols_of_formula[clause.formula].remove(symbol)
+                    yield from self._remove_symbol_from_formula(clause.formula, symbol)
+        self._logic_graph.remove_node(clause)
+
+    def _remove_symbol(self, symbol: Symbol) -> Iterator[Clause]:
+        """Remove the given symbol from the graph."""
+        yield from self._logic_graph.predecessors(symbol)
+        for formula in self._formulas_containing_symbol[symbol]:
+            self._symbols_of_formula[formula].remove(symbol)
+        self._logic_graph.remove_node(symbol)
+
+    def _remove_symbol_from_formula(self, formula: Formula, symbol: Symbol) -> Iterator[Symbol]:
+        """
+        Update the dictionaries and decides whether we also remove the given symbol, if the symbol is not contained in the given formula.
+        """
+        self._formulas_containing_symbol[symbol].remove(formula)
+        if len(self._formulas_containing_symbol[symbol]) <= 1:
+            yield symbol
+
+    # def get_next_subexpression(self) -> Iterator[Tuple[AbstractSyntaxTreeNode, LogicCondition]]:
+    #     """Get the next subexpression together with the node it comes from and start with the largest possible subexpression!"""
+    #     current_size = self.maximum_subexpression_size
+    #     while current_size > 0:
+    #         for ast_node in [c for c, p in self._candidates.items() if self._logic_graph.out_degree(p) >= current_size]:
+    #             if ast_node not in self._candidates:
+    #                 continue
+    #             if current_size > self.maximum_subexpression_size:
+    #                 break
+    #             clauses = self._get_clauses(ast_node)
+    #             for new_operands in combinations(clauses, current_size):
+    #                 yield ast_node, LogicCondition.conjunction_of(new_operands)
+    #                 if ast_node not in self._candidates or current_size > self.maximum_subexpression_size:
+    #                     break
+    #         current_size = min(self.maximum_subexpression_size, current_size - 1)
```

Comment:
Actually no, this was an option where we start with the largest subexpression and then go to the smaller, but this sometimes led to worse results. I was not sure whether this is because we do not consider condition nodes, or whether this is general the case. I was thinking of trying this out after also allowing condition-nodes, but I can remove it and save it somewhere else if you prefer.

Reply:
That would be great, I'm afraid it would only collect dust here and make people wonder what is up with it